### ISSUE TITLE
Deprecate before changing interface of Strong Params

### ISSUE THREAD

**Issue by schneems** [rails/rails#23084]
For Rails 5 strong params was changed to not inherit from hash. I think that's a great idea, however...we didn't deprecate the change (did we?). I think we should add a method missing that will send to the underlying hash and issue a deprecation before we remove the method missing and deprecation in 5.1.

This sudden change bit me on codetriage when upgrading.

cc/ @nateberkopec @sikachu

**Comment by maclover7** [comment]
Inheritance was dropped [here](https://github.com/rails/rails/commit/14a3bd520dd4bbf1247fd3e0071b59c02c115ce0#diff-72f51b082f5c2a99c328f5f5bbbe8ec6), looked back through history, and don't see any deprecation notices. What would the `method_missing` you envision look like? Can take a stab at this.

**Comment by rafaelfranca** [comment]
I don't think it is a good idea. The change was to avoid security issues and adding a `method_missing` would reintroduce them.

What is the problem that you are facing in the upgrade. If it is `NoMethodError` I don't see any problem with the current upgrade path.

**Comment by schneems** [comment]
This broke with no warning while trying to upgrade https://github.com/codetriage/codetriage/commit/4184865ce40c533722983dbc27fa2406b0e45662

**Comment by maclover7** [comment]
I believe that we (at least) have this in the Rails 5.0 upgrade guide.http://edgeguides.rubyonrails.org/5_0_release_notes.html

**Comment by schneems** [comment]
This is going to be a breaking change on just about every upgrading app.

**Comment by schneems** [comment]
@rafaelfranca put that in the warning

```
"You are using StrongParameters as a hash which is potentially unsafe. This behavior will be deprecated removed in Rails 5.1."
```

If people want to be safer they can choose to not use the old API.

**Comment by rafaelfranca** [comment]
Yes, it is expected to be a breaking change. It is not a hard one to fix. You get `NoMethodError`, you change the code, you run the tests.

**Comment by kaspth** [comment]
@schneems When you checked the original discussions around this, did they have anything alluding to why we didn't go with a deprecation cycle?

**Comment by schneems** [comment]
I don't think it was mentioned https://github.com/rails/rails/pull/20868 which is the merged implementation. It was mentioned 35 times in the original un-merged implementation https://github.com/rails/rails/pull/14384

**Comment by rafaelfranca** [comment]
We did soft deprecations in Rails 4.2 https://github.com/rails/rails/pull/16299/files#diff-c909acd0bcbfca7336cca93598e7c65aR1

**Comment by rafaelfranca** [comment]
BTW, what was mentioned in the original issue was to deprecate the `is_a?` check with `Hash` as parameters. The methods were not deprecated because the `Parameters` API was already board enough to have almost all Hash objects that an application use. Of course there are methods that are not in the API yet and I think it is fine, like I said you get the `NoMethodError`, you change the code.

**Comment by schneems** [comment]
It is easy to see where the problem is, but it can also delay upgrading. By making a NoMethodError over a deprecation changing is required to upgrade, instead of optional. I had to open up the Rails source code to see what methods were available to me. Would be really easy to to do something like this


```ruby
  def self.method_missing(method_sym, *args, &block)
    if @parameters.respond_to?(method_sym)
      deprecate("Method #{method_sym} is expecting a Strong parameters to be a hash, instead use #{ self.methods.inspect }")
      @parameters.public_send(method_sym, *args, &block)
    else
      super
    end
  end
end
```

You only get security problems if you're using deprecated methods, you're only getting performance hit if you're using deprecated methods, but it can make the upgrade process more pleasant. 

From @jeremy on the first issue

> Another way is to proxy Parameters#missing_method to the composed HWIA instance, along with a deprecation warning that params will no longer be a Hash in Rails 5.

This is exactly what i'm proposing.

**Comment by nateberkopec** [comment]
It is an extremely far-reaching change and I suspect nearly every app will have to deal with. I did on my relatively trivial [Turbolinks TodoMVC demo](https://github.com/nateberkopec/todomvc-turbolinks/commit/e2a5efb40bdf54aa1b15972036a3c55c2d49b566). It's a good change, and I think we all understand it needs to be done.

I don't have enough experience to really weigh in on the appropriate strategy here, but I think at least the exception could be made more clear. NoMethodError in this case is not immediately edifying as to *why* this is going on - I think I only figured it out because I remembered this change and went back and looked it up.

We could provide a more accurate exception message, perhaps with a link to a guide or explanation. Without even an explanation of the reasons for this change, many people (like I did) will just call `to_unsafe_h` and move on. That would help more than it hurts.

**Comment by rafaelfranca** [comment]
How many methods are having this problem?

**Comment by rafaelfranca** [comment]
If the implications of the usage of the wrong methods is clear in the
deprecation message I don't see why reason why not do it.

**Comment by sikachu** [comment]
This change was to prevent a security issue, since it's pretty fatal if an unfiltered hash get passed to any multi-attributes setter method. Redefining a `method_missing` and sending it to underlying hash really going to create a risk if you somehow ended up getting a plain Ruby `Hash` and not `Parameters` object. It's better safe than sorry.

We were going aggressive on this one due to a risk of the previous implementation. Now that we thought about it more, I think it make sense to relax it a bit.

So, we're going to:

1. Introduce `method_missing` which emits a warning and then pass on the message to the underlying Hash object. This warning will state that this method is not implemented in `Parameters` and will break in 5.1. Some apps will still break, but at least you'll have a warning message.
2. Defines `values` on `Parameters` object, since that seems like a good method to have it on there.

I would like to state that it's okay to submit a PR to add useful missing methods to `Parameters` object. Those set of methods are not something set-in-the-stone, but was implemented there just because there were some test case covering the usage of it. Something like `values` is definitely a good addition.

**Comment by schneems** [comment]
Please take a look at PR #23123. I added the deprecation and a few values methods in to the delegate list.
